README.md
# sopalin

Wipe up spilled milk in Hapi.

A plugin to shutdown on uncaughtExceptions, while allowing in-flight responses to complete. While the server is shutting down,
new requests will be responded to with a `503`.

## Registration options

- `replyHeaders` - Response headers to return post error, while the server is still shutting down. Defaults to `{}`.
- `shutdownTimeout` - Timeout option for Hapi.Server#stop. Defaults to `10000`.
- `lastly(error)` - An optional final callback for clean up. Called after server shutdown. The default behavior is to `process.exit(1)`.

## Usage

```javascript
const Hapi = require('hapi');
const Sopalin = require('sopalin');

const server = new Hapi.Server();

server.register({
    register: Sopalin,
    options: {
        shutdownTimeout: 15000,
        replyHeaders: {
            'x-custom-retry-header': 'true'
        }
    }
}, (error) => {
    //Everything else...
});
```


CHANGELOG.md

### v1.0.1

- Added default `connection: 'close'` header.
- Log to `console.warn` instead of `server.log` after server shutdown.


test/test-sopalin.js
'use strict';

const Test = require('tape');
const Sopalin = require('../index');
const Hapi = require('hapi');
const Boom = require('boom');

Test('test sopalin', (t) => {

    t.test('with lastly', (t) => {
        t.plan(8);

        const server = new Hapi.Server();

        server.connection({ port: 3000 });

        const plugins = [
            {
                register: Sopalin.register,
                options: {
                    replyHeaders: {
                        'a': 'a-header'
                    },
                    lastly: (error) => {
                        t.ok(error, 'error.');
                        t.pass('lastly called.');
                        t.equal(error.message, 'Uncaught error: Something blew up!', 'error passed to lastly.');
                    }
                }
            },
            {
                register: require('good'),
                options: {
                    reporters: {
                        consoleReporter: [
                            {
                                module: 'good-console',
                            },
                            'stdout'
                        ]
                    }
                }
            }
        ];

        server.register(plugins, (error) => {
            t.error(error, 'should not be an error.');

            server.route({
                method: 'GET',
                path: '/',
                handler: function (request, reply) {
                    setImmediate(() => {
                        throw new Error('Something blew up!');
                    });
                }
            });

            server.inject({
                method: 'GET',
                url: '/'
            }, (response) => {
                t.equal(response.statusCode, 500, 'error response.');
                setImmediate(() => {
                    server.inject({
                        method: 'GET',
                        url: '/'
                    }, (response) => {
                        t.equal(response.statusCode, 503, '503 response.');
                        t.equal(response.headers.a, 'a-header', 'replyHeaders passed.');
                        t.equal(response.headers.connection, 'close', 'connection close header passed.');
                    });
                });
            });
        });

    });

    t.test('without lastly', (t) => {
        t.plan(3);

        const server = new Hapi.Server();

        server.connection({ port: 3000 });

        const plugins = [
            {
                register: Sopalin.register
            }
        ];

        const exit = process.exit;

        process.exit = function (n) {
            t.equal(n, 1, 'process.exit(1) called.');
        };

        t.once('end', () => {
            process.exit = exit;
        });

        server.register(plugins, (error) => {
            t.error(error, 'should not be an error.');

            server.route({
                method: 'GET',
                path: '/',
                handler: function (request, reply) {
                    setImmediate(() => {
                        throw new Error('Something blew up!');
                    });
                }
            });

            server.inject({
                method: 'GET',
                url: '/'
            }, (response) => {
                t.equal(response.statusCode, 500, 'error response.');
            });
        });

    });

    t.test('non domain-error', (t) => {
        t.plan(3);

        const server = new Hapi.Server();

        server.connection({ port: 3000 });

        const plugins = [
            {
                register: Sopalin.register
            }
        ];

        const exit = process.exit;
        let exitCalled = false;

        process.exit = function (n) {
            exitCalled = true;
        };

        t.once('end', () => {
            process.exit = exit;
        });

        server.register(plugins, (error) => {
            t.error(error, 'should not be an error.');

            server.route({
                method: 'GET',
                path: '/',
                handler: function (request, reply) {
                    reply(Boom.badImplementation());
                }
            });

            server.inject({
                method: 'GET',
                url: '/'
            }, (response) => {
                t.equal(response.statusCode, 500, 'error response.');
                setTimeout(() => t.ok(!exitCalled, 'process.exit not called.'), 100);
            });
        });

    });

});


index.js
'use strict';

const Pkg = require('./package.json');
const Hoek = require('hoek');
const Joi = require('joi');
const Boom = require('boom');

const internals = {
    info: {
        name: Pkg.name,
        version: Pkg.version
    },

    defaults: {
        shutdownTimeout: 10000,
        replyHeaders: {}
    }
};

const schema = Joi.object({
    lastly: Joi.func().allow(null),
    shutdownTimeout: Joi.number().default(internals.defaults.shutdownTimeout),
    replyHeaders: Joi.object().default(internals.defaults.replyHeaders)
});

const plugin = {
    register(server, options, next) {
        let shuttingDown = false;

        const validation = Joi.validate(options, schema);

        Hoek.assert(!validation.error, validation.error);

        const { lastly, shutdownTimeout, replyHeaders } = validation.value;

        const errorHandler = function (request, error) {
            if (error.domainThrown) {
                shuttingDown = true;

                server.log(['warn'], 'shutting down.');

                server.root.stop({ timeout: shutdownTimeout }, () => {
                    server.removeListener('request-error', errorHandler);

                    if (lastly) {
                        lastly(error);
                        return;
                    }

                    console.warn(Pkg.name, ': application process exit.');
                    process.exit(1);
                });
            }
        };

        server.on('request-error', errorHandler);

        server.ext('onRequest', (request, reply) => {
            if (shuttingDown) {
                request.log(['warn'], '503 response while shutting down.');

                const errorResponse = Boom.serverUnavailable();

                errorResponse.output.headers = Object.assign({ connection: 'close' }, replyHeaders);

                reply(errorResponse);
                return;
            }

            reply.continue();
        });

        next();
    }
};

plugin.register.attributes = internals.info;

module.exports = plugin;
